GHCN.py

```diff
@@ -0,0 +1,1044 @@
+            count += 8 # here each month has 31 days (padded with missing values)
+        if zz != z+31: raise ParseError, 'Line has {:d} values instead of 31:\n {:s}'.format(zz-z,line)  
+         # increment counter
+        z += 31
+    #if existflag==False:tlen=0
+    #if z < tlen: raise ParseError, 'Reached end of file before specified end date: {:s}'.format(self.filename)    
+    f.close() # close again
+    # return array
+    return data
+  
+
+## class that defines variable properties (specifics are implemented in children)
+class VarDef(RecordClass):
+  # variable specific
+  name        = '' # full variable name (used in source files)
+  atts        = None # dictionary with PyGeoData variable attributes
+  prefix      = '' # file prefix for source file name (used with station ID)
+  fileext     = '.dly' # file name extension (used for source files)
+  # type specific
+  datatype    = '' # defined in child class; type of source file
+  recordunits = '' # units used for data  (in source files)
+  units       = '' # units after conversion
+  scalefactor = 1 # constant scaling factor for conversion 
+  offset      = 0 # constant offset for conversion
+  dtype       = 'float32' # data type used for data in source files
+  encoding    = 'UTF-8' # file encoding (used in source files)
+  missing     = '' # string indicating missing value
+  mflag        = '' # legal data flags (case sensitive)
+  qflag        = ''
+  sflag        = ''
+  varmin      = 0. # smallest allowed value in data
+  varmax      = 0. # largest allowed value in data
+  # inferred variables
+  variable    = '' # alias for name
+  filepath    = '' # inferred from prefix
+  
+  def __init__(self, **kwargs):
+    super(VarDef,self).__init__(**kwargs)
+    self.variable = self.name
+    self.filepath = '{0:s}{1:s}'.format('{:s}',self.fileext)
+    
+  def convert(self, data): 
+    if self.scalefactor != 1: data *= self.scalefactor;
+    if self.offset != 0: data += self.offset
+    return data
+  
+  def getKWargs(self, *args):
+    ''' Return a dictionary with the specified arguments and their values '''
+    if len(args) == 0: 
+      args = ['variable', 'units', 'varmin', 'varmax', 'missing', 'mflag', 'qflag', 'sflag', 'dtype', 'encoding']
+    kwargs = dict()
+    for arg in args:
+      kwargs[arg] = getattr(self,arg)
+    return kwargs
+  
+# definition for precipitation files
+class PrecipDef(VarDef):
+#   recordunits = 'mm' # actually 'mm/day' but reported as 'mm'
+#   units       = 'kg/m^2/s' # units after scaling (SI)
+  scalefactor = .1/(24.*60.*60.) # convert from mm/day to mm/s 
+  offset      = 0 
+  datatype    = 'precip'
+  title       = 'GHCN Precipitation Records'
+  missing     = '-9999' # string indicating missing value (apparently not all have 'M'...)
+  MFLAG1      = ' BDHKLOPTW' # legal data flags (case sensitive; 'M' for missing should be screened earlier)
+  QFLAG1      = ' DGIKLMNORSTWXZ'
+  SFLAG1      = ' 067AaBbCEFGHIKMNQRrSsTUuWXZz'

```

Comment:
These flags are in the ghcn daily station record but they are currently not read at all since I only use those columns that contain numbers.
I leave them here just in case of future use. I will be able to use data with a specific flag if I read those corresponding columns from the ASCII files.


> 在 2016年7月24日，14:30，"Andre R. Erler" <notifications@github.com> 写道：
> 
> In GHCN.py:
> 
> > +    for arg in args:
> > +      kwargs[arg] = getattr(self,arg)
> > +    return kwargs
> > +  
> > +# definition for precipitation files
> > +class PrecipDef(VarDef):
> > +#   recordunits = 'mm' # actually 'mm/day' but reported as 'mm'
> > +#   units       = 'kg/m^2/s' # units after scaling (SI)
> > +  scalefactor = .1/(24.*60.*60.) # convert from mm/day to mm/s 
> > +  offset      = 0 
> > +  datatype    = 'precip'
> > +  title       = 'GHCN Precipitation Records'
> > +  missing     = '-9999' # string indicating missing value (apparently not all have 'M'...)
> > +  MFLAG1      = ' BDHKLOPTW' # legal data flags (case sensitive; 'M' for missing should be screened earlier)
> > +  QFLAG1      = ' DGIKLMNORSTWXZ'
> > +  SFLAG1      = ' 067AaBbCEFGHIKMNQRrSsTUuWXZz'
> what are these flags?
> 
> ―
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>